### DISCUSSION TITLE
Mimir queriers OOMKilling

### DISCUSSION THREAD

**Discussion by LasseHels** [grafana/mimir#7365] [Help and support] [answered]
Hello!

Our Mimir queriers regularly spike in memory usage, and we occasionally get periods of 30-60 minutes where queriers will continuously OOMKill:
![Screenshot 2024-02-12 at 12 43 14](https://github.com/grafana/mimir/assets/121869045/01237c0d-6fc0-494d-87aa-18eec25dd286)

Our platform is used by a wide variety of users, and we initially believed the spiky memory usage was due to certain user query patterns. However, this past Sunday (February 11th), we saw another instance of spiky memory usage. We would expect no or few users to be running queries outside of work hours.

Under normal operations, our queriers consume ~1GB of memory:
![Screenshot 2024-02-12 at 12 45 02](https://github.com/grafana/mimir/assets/121869045/2fee932f-c894-4f40-80b8-5bf8b3408a33)

We're having a hard time finding metrics and/or log lines that correlate with the memory spikes. We would appreciate any help in narrowing down where the excessive memory usage is coming from.

This is our querier configuration:
```yaml
      - args:
        - -auth.multitenancy-enabled=false
        - -auth.no-auth-tenant=fake
        - -blocks-storage.azure.container-name=mimir
        - -blocks-storage.bucket-store.metadata-cache.backend=memcached
        - -blocks-storage.bucket-store.metadata-cache.memcached.addresses=dnssrvnoa+memcached-metadata.mimir.svc.cluster.local.:11211
        - -blocks-storage.bucket-store.metadata-cache.memcached.max-async-concurrency=50
        - -blocks-storage.bucket-store.metadata-cache.memcached.max-item-size=1048576
        - -blocks-storage.bucket-store.sync-dir=/data/tsdb
        - -blocks-storage.bucket-store.sync-interval=15m
        - -common.storage.azure.account-name=pensievemimirprodwe01
        - -common.storage.backend=azure
        - -config.expand-env=true
        - -config.file=/etc/mimir/config/mimir.yaml
        - -distributor.health-check-ingesters=true
        - -ingester.ring.heartbeat-timeout=10m
        - -ingester.ring.prefix=
        - -ingester.ring.replication-factor=3
        - -ingester.ring.store=memberlist
        - -ingester.ring.zone-awareness-enabled=true
        - -mem-ballast-size-bytes=268435456
        - -memberlist.bind-port=7947
        - -memberlist.join=gossip-ring.mimir.svc.cluster.local:7947
        - -querier.cardinality-analysis-enabled=true
        - -querier.frontend-client.grpc-max-send-msg-size=419430400
        - -querier.max-concurrent=100
        - -querier.max-outstanding-requests-per-tenant=300
        - -querier.max-partial-query-length=816h
        - -querier.max-query-lookback=64w
        - -querier.query-ingesters-within=13h
        - -querier.query-store-after=12h
        - -query-scheduler.ring.prefix=
        - -query-scheduler.ring.store=memberlist
        - -query-scheduler.service-discovery-mode=ring
        - -runtime-config.file=/etc/mimir/overrides.yaml
        - -server.grpc.keepalive.min-time-between-pings=10s
        - -server.grpc.keepalive.ping-without-stream-allowed=true
        - -server.http-listen-port=8080
        - -server.http-write-timeout=2m
        - -store-gateway.sharding-ring.prefix=multi-zone/
        - -store-gateway.sharding-ring.replication-factor=3
        - -store-gateway.sharding-ring.store=memberlist
        - -store-gateway.sharding-ring.zone-awareness-enabled=true
        - -target=querier
```

**Comment by pstibrany** [comment, marked as answer]
Hello. If your queriers are configured with activity tracker enabled (it is enabled by default, but requires `emptyDir` volume to store data -- if you use Helm or our Jsonnet, this should be the case), then on next restart queriers will log all ongoing queries that were running when querier crashed. This can be used to find out which queries caused OOM.

> We would expect no or few users to be running queries outside of work hours.

It can be a single query causing OOMs. Once you understand which query it is and what exactly is happening, you can protect queriers by placing limits on the queries.

**Reply by pstibrany** [reply]
In your configuration, some config flags stand out:

```
        - -querier.max-concurrent=100
        - -querier.max-partial-query-length=816h
        - -querier.max-query-lookback=64w
```

These have potential to use extra memory.

**Reply by LasseHels** [reply]
@pstibrany Thanks for the swift reply! Your answer gives us a lot of leads to chase. We will investigate!

**Reply by LasseHels** [reply]
@pstibrany We've now had a chance to run queriers with a volume attached for a few days. Note that we didn't change any other settings, mainly because we want _some_ amount of OOMKills to happen so we can inspect queries from the `metrics-activity.log` file.

Our goal is to try to narrow down the queries that are causing our queriers (and ruler queriers) to OOMKill.

Our methodology thus far has been to use the `fetched_chunk_bytes` and `fetched_index_bytes` labels emitted by the query frontend after processing a query. It is our understanding that the more bytes a given query fetches, the more that query contributes to the querier's memory usage when the querier processes the query.

This methodology has proved dodgy, and we are finding that queries from the `metrics-activity.log` file don't fetch an extraordinary amount of data.

Some examples:
* `sum by (le, pod) (rate(cortex_request_duration_seconds_bucket{job=~"(mimir)/((distributor.*|cortex|mimir|mimir-write.*))",k8s_cluster=~"telemetry-metrics-prod-we-1",route=~"/distributor.Distributor/Push|/httpgrpc.*|api_(v1|prom)_push|otlp_v1_metrics"}[1m]))` this query fetched 0.0217 GiB of data.
* `idc_mem_total{product_id="idc-cap-prodws1"}` this query fetched 0.00000534 GiB of data.
* `max by (container, app, uri) (http_server_requests_seconds{app="athena-wnd-ui",k8s_cluster="athena-prod-westeurope-1",quantile="0.95"})` this query fetched 0 GiB of data.

For reference, our queriers run with a memory limit of 50 GiB. Our ruler queriers run with a memory limit of 10 GiB.

The way we're determining how much data a query fetches is by running the query with a time range of one hour, and then inspecting the bytes fetched with this Loki query:
```
(
    sum(sum_over_time({product_id="telemetry", namespace="mimir", component="query-frontend"} |= "sum by (le, pod) (rate(cortex_request_duration_seconds_bucket{job=~" | logfmt | unwrap fetched_chunk_bytes[$__auto]))
    +
    sum(sum_over_time({product_id="telemetry", namespace="mimir", component="query-frontend"} |= "sum by (le, pod) (rate(cortex_request_duration_seconds_bucket{job=~" | logfmt | unwrap fetched_index_bytes[$__auto]))
)
/ 1024 / 1024 / 1024
```

Is our understanding of the relationship between fetched bytes and OOMKills correct? Is there a more accurate signal we can measure?

**Reply by pstibrany** [reply]
> This methodology has proved dodgy, and we are finding that queries from the metrics-activity.log file don't fetch an extraordinary amount of data.

It's not clear to me if you use the file correctly. You're not supposed read the file by yourself -- queriers write *ALL* the queries into the file, while queries are executed.

The file is *only* useful when querier crashes. Then on next restart querier will read the file and log all queries that were running during the crash. If same query causes multiple crashes, this is typically easy to see from the log messages.

> Our methodology thus far has been to use the fetched_chunk_bytes and fetched_index_bytes labels emitted by the query frontend after processing a query.

Yes, there is strong correlation between OOMs and fetched chunk/index bytes. One thing to keep in mind is that message logged by query-frontend is aggregation over all subqueries. If single query from client is split (by time) and/or sharded (for shardable queries) into let's say 100 small queries executed by querier, they can together fetch large number of bytes, but each query may still work just fine. I would therefore suggest looking at `split_queries` and `sharded_queries` too, and perhaps work with per-query average numbers.

**Reply by pstibrany** [reply]
> Yes, there is strong correlation between OOMs and fetched chunk/index bytes

Let me double check with my colleagues on this. There was a lot of work done on read path to keep memory usage under control, so I may be wrong here.

**Reply by LasseHels** [reply]
@pstibrany

> It's not clear to me if you use the file correctly. You're not supposed read the file by yourself -- queriers write ALL the queries into the file, while queries are executed.
>
> The file is only useful when querier crashes. Then on next restart querier will read the file and log all queries that were running during the crash. If same query causes multiple crashes, this is typically easy to see from the log messages.

Apologies, I should have been more specific. We're not reading directly from the file. We're reading queries from the log lines emitted by the querier on startup after a crash.

**Reply by dimitarvdimitrov** [reply]
What @pstibrany wrote it 100% correct. There is a strong correlation between fetched chunks/index and OOMs. Note that not all queries will cause OOMs, there could be a select few in the activity log of a querier which were the culprits. It's also possible to just have too many inflight queries. For example we prefer having the max concurrency of 8 and prefer having more querier replicas. 100 seems like a lot https://github.com/grafana/mimir/blob/7771731dd7e4c3b00d5ab09127952058f69f7be7/operations/mimir/config.libsonnet#L43


Analyzing the queries in the activity log after a few crashes should show a pattern. Perhaps correlating the number of inflight requests at the queriers with OOMs might help prove or disprove the theory for querier_max_concurrent - you can look at `sum(rate(cortex_querier_request_duration_seconds_sum{job="querier"}[1m]))` to see the average number of concurrent queries (note `rate(_sum)`, not `rate(_count)`)

Some things to look out for when correlating queries in the activity log and other occurrences of the same query
* is the time range the same? Recent versions of the "Mimir / Slow Queries" dashboard have a more human-readable format for the query time span
* were the selectors the same? In your logQL query you include all queries that select `cortex_request_duration_seconds_bucket`. In this particular example, this can be all ingesters or all overrides exporters. The difference between those two sets is probably significant.

**Reply by LasseHels** [reply]
@dimitarvdimitrov Thanks for your reply! We will take a look at the suggested items.

**Reply by LasseHels** [reply]
We've lowered `querier.max-concurrent` from `100` to `8`. This has had _some_ positive effect, but we still see regular OOMKills.

We are looking at different Mimir configuration options, and were wondering whether modifying our sharding configuration could have some effect on memory usage?

This is our current sharding configuration
```yaml
        - -query-frontend.parallelize-shardable-queries=true
        - -query-frontend.query-sharding-max-sharded-queries=128
        - -query-frontend.query-sharding-target-series-per-shard=2500
        - -query-frontend.query-sharding-total-shards=16
```

Specifically, we are thinking about lowering `-query-frontend.query-sharding-target-series-per-shard`, but our understanding of how this configuration option works is fairly limited.

We are also looking at the `-querier.max-samples` configuration option. Right now, we have it set to the default of `50000000`. We would expect there to be a correlation between the amount of samples that a given query fetches and the memory usage incurred by that query. Is that true? If so, lowering `-querier.max-samples` could prevent costly queries from being ran at all.

**Reply by dimitarvdimitrov** [reply]
> Specifically, we are thinking about lowering `-query-frontend.query-sharding-target-series-per-shard`, but our understanding of how this configuration option works is fairly limited.

this flag helps when the same query has been ran previously over a different time range. The feature uses the actually processed series in the previous request and use them to estimate the shards for the current query.

> We would expect there to be a correlation between the amount of samples that a given query fetches and the memory usage incurred by that query. Is that true?

that's true. Also consider lowering the limits `-querier.max-fetched-series-per-query` and `-querier.max-fetched-chunks-per-query`, which can also be configured on a per-tenant basis. Those limits are applied earlier in the query path and can be hit before the querier has to load everything in memory to run the promQL engine. So they have somewhat better protection against OOMs.

**Reply by LasseHels** [reply]
For the record, we were able to _mostly_ solve this issue by changing only `-querier.max-concurrent` back down to `8` for both our regular queriers and ruler queriers. We still occasionally see OOMKills, but the rate has been reduced to the point where it's not a problem in practice.

Thanks for the help @dimitarvdimitrov and @pstibrany ❤️.

**Reply by pstibrany** [reply]
That's great to hear!
